Messaging: clarify network attributes usage on common semconv

lmolkova opened the pull request:
## Changes

Addresses messaging portion of #690

Removes `network.transport|type` attributes from general messaging semantic conventions.

Based on the findings from Java instrumentation libraries, the only messaging instrumentation that populates network attributes is RabbitMQ (which also sets them in native .NET instrumentation).  

This PR clarifies that `network.server.address|port` should be only set when they represent a specific broker instance (when individual instances are visible from the application side) and recommends to use them on the RabbitMQ (as one of the systems that's usually self-hosted)

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* ~~schema-next.yaml~~

pyohannes commented:
From the messaging point of view, I'd agree to merge those changes.

lmolkova commented on model/trace/database.yaml:
```diff
@@ -36,16 +36,6 @@ groups:
         tag: connection-level
         requirement_level:
           conditionally_required: If using a port other than the default port for this DBMS and if `server.address` is set.
-      - ref: network.peer.address
```
TODO: also update examples

joaopgrassi replied:
Will you update them as part of this PR or as a follow up?

lmolkova replied:
In this PR, I'd like to get a blessing from DB WG first anyway.

joaopgrassi commented on .chloggen/698.yaml:
```diff
@@ -0,0 +1,7 @@
+change_type: breaking
+
+component: messaging
```
```suggestion
component: messaging, db
```

Either that (which I'm not sure it's a good idea..) or split the db changes into a separated PR.

I tried doing it the former locally, and it renders like this:

```
### 🛑 Breaking changes 🛑

- `messaging, db`: test (#123)
```

lmolkova replied:
ah, good point! 

I'll bring it up on db semconv and if there is support for this PR, I'll split this Pr in two. Otherwise, I'll close it.

I feel we should have the same decision for both cases and it wouldn't make sense to change messaging, but keep DB stuff.
Making it as draft for now

trask reviewed (commented):
it would be great to look over the existing Java database instrumentation to see how many places where we're (usefully/correctly?) capturing network attributes

lmolkova commented:
Looking into java instrumentation repo:

Messaging systems:
- report only network attributes:
  - rabbitmq, but reporting network stuff is not straightforward involving [thread locals, passing things through otel context](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/778fc3eb89000f8900f1591ebbc5987056bbacb6/instrumentation/rabbitmq-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rabbitmq/RabbitChannelInstrumentation.java#L116)
- report server and network attributes:
  - aws sns/sqs server and only `network.protocol.verion` out of all network attributes
- report server attributes only:
  - pulsar
- report no server/network attributes at all: kafka, rocketmq

Databases:
- report only network:
  - cassandra
  - Redisson
- report only server: 
  - jdbc
  - mongo
  - r2
  - vertxsql
  - elastic rest
- report both, server and network attributes:  
  - dynamodb (also reports `network.protocol.verion`, but no other network)
  - jedis (depends on the version)
  - lettuce
  - vertx redis
  - elastic-transport
  - opensearch
- report nothing (no network or server):  geodedb, Rediscala, Spymemcached

All instrumentations that report network, report only `network.peer.address|port` and some report `network.type`. I could not find any that reports protocol name or transport.

I probably missed something, but I think there is enough evidence. 

- `server.*` or `network` attributes cannot be required on DB or messaging
- I *think* instrumentations that support network, but don't report `server.address` should try by calling `getHostName()` on `Inet*Address` they already have
- I still want to remove `network` from general db, we keep it for redis, cassandra, and elastic.
- I still want to remove `network` from general messaging, but want to keep them for rabbitmq

pyohannes commented:
> Looking into java instrumentation repo:

Thanks @lmolkova for this research.

To add one more datapoint, the [.NET RabbitMQ instrumentation](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1261) populates both network and server attributes (if they are available):
* `network.peer.address`
* `network.peer.port`
* `network.local.address`
* `network.local.port`
* `network.type`
* `network.protocol.name`
* `network.protocol.version`
* `server.address`
* `server.port`

lmolkova commented:
Based on the discussion in Messaging SIG, breaking this one down into two PRs. This one contains messaging portion only. 

Will send a separate one for DBs.

lmolkova commented:
Based on the database discussion on similar PR https://github.com/open-telemetry/semantic-conventions/pull/768#issuecomment-1986534133, I updated this one to be consistent with databases:

- removed `network.transport|type`
- kept `network.peer.*` in common, but clarified they apply for self-hosted systems where they make sense

lmolkova commented on docs/messaging/gcp-pubsub.md:
```diff
@@ -13,7 +13,7 @@ The Semantic Conventions for [Google Cloud Pub/Sub](https://cloud.google.com/pub
 ## Span attributes
 
 For Google Cloud Pub/Sub, the following additional attributes are defined:
-<!-- semconv messaging.gcp_pubsub(full,tag=tech-specific-gcp-pubsub) -->
```
cosmetic change that simplifies tag usage a bit

pyohannes commented on docs/messaging/messaging-spans.md:
```diff
@@ -348,23 +342,13 @@ size should be used.
-**[16]:** `network.protocol.version` refers to the version of the protocol used and might be different from the protocol client's version. If the HTTP client has a version of `0.27.2`, but sends HTTP version `1.1`, this attribute should be set to `1.1`.
-
-**[17]:** The value SHOULD be normalized to lowercase.
+**[14]:** Network peer address and port are important when the application interacts with individual intermediary nodes directly, for example, when messaging intermediary is self-hosted. Fully managed messaging systems that don't expose information about individual nodes SHOULD NOT set this attribute. Semantic conventions for individual messaging systems SHOULD document whether `network.peer.*` a ...
```
Can client instrumentation know whether it's interacting with a self-hosted or fully-managed?

If for a given system possibly both self-hosted and fully-managed offerings exist, what should client instrumentation do? As in may cases it can likely not know whether a given address refers to a self-hosted or managed instance.

lmolkova replied:
The intention is to let individual systems decide if/how/when `network.peer.*` should be collected. I rephased it a bit to make it more clear and update this sentence to "Instrumentations for messaging systems that don't provide self-hosted mode SHOULD NOT set this attribute." LMK what you think.

lmolkova replied:
@pyohannes could you please take a look and lmk if you have any objections?

pyohannes replied:
This looks good.

trask commented on model/trace/messaging.yaml:
```diff
@@ -80,28 +80,55 @@ groups:
           conditionally_required: When available.
-        tag: tech-specific-rabbitmq
+        tag: tech-specific
+      - ref: network.peer.address
+        note: >
+          If an operation involved multiple network calls (for example retries),
+          the address of the last contacted node SHOULD be used.
+        tag: tech-specific
+      - ref: network.peer.port
+        tag: tech-specific
```
are these needed since it extends `messaging.tech_specific.network.attributes`?

trask commented on model/trace/messaging.yaml:
```diff
@@ -80,28 +80,55 @@ groups:
+        note: >
+          Semantic conventions for individual messaging systems SHOULD document whether `network.peer.*` attributes are applicable.
+
+          Network peer address and port are important when the application interacts with individual intermediary nodes directly
```
```suggestion
          Network peer address and port are important when the application interacts with individual intermediary nodes directly.
```

trask commented on docs/messaging/kafka.md:
```diff
@@ -24,10 +24,9 @@ described on this page.
+<!-- semconv messaging.kafka(full,tag=tech-specific) -->
 | Attribute  | Type | Description  | Examples  | Requirement Level |
 |---|---|---|---|---|
-| [`messaging.destination.partition.id`](../attributes-registry/messaging.md) | string | "String representation of the partition id the message  (or batch) is sent to or received from."" | `1` | Recommended |
```
what happened here?

joaopgrassi replied:
I think it may be a left over from https://github.com/open-telemetry/semantic-conventions/pull/798/files

lmolkova replied:
Great catch, thanks, fixed

trask commented on docs/messaging/rabbitmq.md:
```diff
@@ -17,11 +17,15 @@ described on this page.
 | [`messaging.rabbitmq.destination.routing_key`](../attributes-registry/messaging.md) | string | RabbitMQ message routing key. | `myKey` | Conditionally Required: If not empty. |
 | [`messaging.rabbitmq.message.delivery_tag`](../attributes-registry/messaging.md) | int | RabbitMQ message delivery tag | `123` | Conditionally Required: When available. |
+| [`network.peer.address`](../attributes-registry/network.md) | string | Peer address of the messaging intermediary node where the operation was performed. [1] | `10.1.2.80`; `/tmp/my.sock` | Recommended |
+| [`network.peer.port`](../attributes-registry/network.md) | int | Peer port number of the network connection. | `65123` | Recommended |
```
Maybe worth overriding description to match `network.peer.address`, e.g. “Peer port of the messaging intermediary node where the operation was performed”
